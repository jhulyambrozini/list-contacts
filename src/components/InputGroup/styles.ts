import styled from 'styled-components'
import { colors } from '../../styles'

export const InputGroupContainer = styled.div`
  width: 100%;
  position: relative;
  margin-bottom: 4rem;

  label {
    color: ${colors.grey01};
    font-size: 1.8rem;
    position: absolute;
    top: -16px;
    left: 12px;
    background-color: #fff;
    padding: 4px 8px;
  }

  input {
    width: 100%;
    border: none;
    border: 1px solid ${colors.grey02};
    border-radius: 5px;
    padding: 1.4rem;
  }

  @media screen and (max-width: 426px) {
    label {
      font-size: 1.6rem;
    }
  }
`
