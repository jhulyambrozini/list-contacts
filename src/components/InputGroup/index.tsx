import { ChangeEventHandler, FocusEventHandler } from 'react'
import * as Style from './styles'

export type InputGroupProps = {
  id: string
  type: string
  label: string
  value?: string
  onChange?: ChangeEventHandler<HTMLInputElement>
  onBlur?: FocusEventHandler<HTMLInputElement>
  erroMessage?: string | false | undefined
  isSuccess?: boolean
}

const InputGroup = ({
  id,
  type,
  label,
  value,
  onChange,
  onBlur,
  erroMessage,
  isSuccess
}: InputGroupProps) => (
  <Style.InputGroupContainer>
    <label htmlFor={id}>{label}</label>
    <input
      type={type}
      id={id}
      value={value}
      onChange={onChange}
      onBlur={onBlur}
      className={(isSuccess ? 'success' : '') || (erroMessage ? 'error' : '')}
    />
    <span>{erroMessage}</span>
  </Style.InputGroupContainer>
)

export default InputGroup
