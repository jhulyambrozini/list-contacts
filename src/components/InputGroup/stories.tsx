import { Meta } from '@storybook/react'
import InputGroup, { InputGroupProps } from '.'

const meta = {
  title: 'components/InputGroup',
  component: InputGroup,
  tags: ['autodocs'],
  parameters: {
    layout: 'centered'
  },

  args: {
    id: 'email',
    type: 'email',
    label: 'EMAIL'
  },
  argTypes: {
    erroMessage: { control: { disable: true } },
    isSuccess: { control: { disable: true } },
    onBlur: { control: { disable: true } },
    onChange: { control: { disable: true } }
  }
} satisfies Meta<typeof InputGroup>

export default meta

export const Template = (args: InputGroupProps) => {
  return <InputGroup {...args} />
}

export const WithError = (args: InputGroupProps) => {
  return (
    <InputGroup
      {...args}
      erroMessage="Email invÃ¡lido"
      value="kamillygmail.com"
    />
  )
}

export const WithSuccess = (args: InputGroupProps) => {
  return <InputGroup {...args} isSuccess value="kamilly@gmail.com" />
}
