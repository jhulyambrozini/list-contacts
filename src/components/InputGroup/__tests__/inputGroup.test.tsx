import InputGroup from '..'
import '@testing-library/jest-dom'
import { fireEvent, render, screen } from '@testing-library/react'

const propsInputGroup = {
  id: 'name',
  type: 'text',
  label: 'NOME'
}

describe('<InputGroup />', () => {
  it('deve renderizar corretamente sem erro ou sucesso', () => {
    const { container } = render(<InputGroup {...propsInputGroup} />)

    const label = screen.getByText('NOME')
    const input = screen.getByRole('textbox', { name: /NOME/i })

    expect(label).toBeInTheDocument()
    expect(input).toBeInTheDocument()

    expect(container.firstChild).toMatchSnapshot()
  })

  it('deve renderizar com erro quando receber uma mensagem de erro', () => {
    const { container } = render(
      <InputGroup {...propsInputGroup} erroMessage="campo obrigatório" />
    )

    const spanErro = screen.getByText('campo obrigatório')
    const input = screen.getByRole('textbox', { name: /NOME/i })

    expect(input.className).toBe('error')
    expect(spanErro).toBeInTheDocument()
    expect(container.firstChild).toMatchSnapshot()
  })

  it(' deve renderizar no estado sucesso quando isSucess for true', () => {
    const { container } = render(<InputGroup {...propsInputGroup} isSuccess />)

    const input = screen.getByRole('textbox', { name: /NOME/i })
    expect(input.className).toBe('success')
    expect(container.firstChild).toMatchSnapshot()
  })

  it(' deve receber corretamente o evento de change', () => {
    render(<InputGroup {...propsInputGroup} />)

    const input = screen.getByRole('textbox', { name: /NOME/i })

    fireEvent.change(input, {
      target: {
        value: 'adicioanndo um texto'
      }
    })
    expect(input).toHaveValue('adicioanndo um texto')
  })
})
