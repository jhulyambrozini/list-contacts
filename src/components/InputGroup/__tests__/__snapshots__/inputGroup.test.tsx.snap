// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<InputGroup /> > should render correctly without error or success 1`] = `
<div
  aria-label="Input group"
  class="sc-bcPKhP eBMfoZ"
>
  <label
    for="name"
  >
    NOME
  </label>
  <input
    class=""
    id="name"
    type="text"
    value=""
  />
  <span />
</div>
`;

exports[`<InputGroup /> > should render in success state when isSucess is true 1`] = `
<div
  aria-label="Input group"
  class="sc-bcPKhP eBMfoZ"
>
  <label
    for="name"
  >
    NOME
  </label>
  <input
    class="success"
    id="name"
    type="text"
    value=""
  />
  <span />
</div>
`;

exports[`<InputGroup /> > should render with error when getting an error message 1`] = `
<div
  aria-label="Input group"
  class="sc-bcPKhP eBMfoZ"
>
  <label
    for="name"
  >
    NOME
  </label>
  <input
    class="error"
    id="name"
    type="text"
    value=""
  />
  <span>
    campo obrigat√≥rio
  </span>
</div>
`;
